server/auth/auth.go

```diff
@@ -691,19 +823,128 @@ func (auth *AuthManager) Notify(acctID account.AccountID, msg *msgjson.Message,
 	auth.SendWhenConnected(acctID, msg, timeout, missedFn)
 }
 
-// Penalize signals that a user has broken a rule of community conduct, and that
-// their account should be penalized.
-func (auth *AuthManager) Penalize(user account.AccountID, rule account.Rule, extraDetails string) error {
+func (auth *AuthManager) registerMatchOutcome(user account.AccountID, misstep NoActionStep, refTime time.Time) int32 {
+	violation := misstep.Violation()
+
+	auth.violationMtx.Lock()
+	defer auth.violationMtx.Unlock()
+	outcomes, found := auth.matchOutcomes[user]
+	if !found {
+		outcomes = newLatest(scoringMatchLimit)
+		auth.matchOutcomes[user] = outcomes
+	}
+	outcomes.add(&stampedFlag{
+		flag: int64(violation),
+		time: encode.UnixMilli(refTime),
+	})
+
+	// Update the match inaction violations in the combined violation counter.
+	counts, found := auth.violationCounts[user]
+	if !found {
+		counts = make(violationCounter)
+		auth.violationCounts[user] = counts
+	}
+	for v, count := range outcomes.bin() {
+		counts[Violation(v)] = uint32(count)
+	}
+
+	score := counts.score()
```

Comment:
I think there might be a bug here. I also think we should get rid of the `violationCounts` field and the `violationCounter` type. Let me explain. 

Let's say the we have `scoringMatchLimit = 3`, and our `latest` looks like `{success, success, noSwapAsTakerScore}` (newer first). Our score would be 11 - 2 = 9. Then we get a tally for another `SwapSuccess`. Once added, our `latest` should look like `{success, success, success}`, but since we're not deleting the `noSwapAsTakerScore` entry in the `auth.violationCounts[user]` map, only overwriting the `success` entry, our `score` would be `11 - 3 = 8` instead of the expected `-3`. 

Even if that there wasn't a bug, I would be a little uncomfortable with keeping the `auth.violationCounts` dict, as it doesn't hold any data that isn't already in `latest`, and keeping both in sync is error-prone. Can we just implement the `score` method on the `latest` type instead?

Reply:
You're right, that's a bug.  I kept the violationCounts map around after adding the two outcomes maps as a tool to integrate from both (note that getting user score required both maps, so the idea is to update the specific outcome map *and* the violation counter so that when other types of outcomes are added it's only necessary to update those kinds of outcome's violations).  Clearly error prone so I'll look at changing it.